redis/src/aio/connection.rs

```diff
@@ -256,109 +258,350 @@ where
     }
 }
 
-/// Represents a `PubSub` connection.
-pub struct PubSub<C = Pin<Box<dyn AsyncStream + Send + Sync>>>(Connection<C>);
+pin_project! {
+    /// Represents a pubsub connection.
+    #[project = PubSubProj]
+    pub struct PubSub<C = Pin<Box<dyn AsyncStream + Send + Sync>>> {
+        #[pin]
+        framed: Framed<C, ValueCodec>,
+        queue: VecDeque<Msg>,
+    }
+}
 
-/// Represents a `Monitor` connection.
-pub struct Monitor<C = Pin<Box<dyn AsyncStream + Send + Sync>>>(Connection<C>);
+fn next_response_err() -> RedisError {
+    RedisError::from((
+        ErrorKind::ResponseDropped,
+        "Response was dropped with stream",
+    ))
+}
 
 impl<C> PubSub<C>
 where
     C: Unpin + AsyncRead + AsyncWrite + Send,
 {
-    fn new(con: Connection<C>) -> Self {
-        Self(con)
+    fn new(con: Connection<C>) -> PubSub<C> {
+        let framed = ValueCodec::default().framed(con.con);
+
+        PubSub {
+            framed,
+            queue: Default::default(),
+        }
+    }
+
+    /// Splits the [`PubSub`] into a [`SplitPubSubSink`] and [`SplitPubSubStream`].
+    pub fn split(self) -> (SplitPubSubSink<C>, SplitPubSubStream<C>) {
+        SplitPubSubSink::new(self.framed)
```

Comment:
I'll have to think about that.

Reply:
> Having the channel tx/rx in a single struct seems inelegant.

I agree with you, but IMO so if writing the same logic twice :)
Maybe it's possible to write some shared logic using traits, but without thinking about it very much, it sounds like you'll need to write some code to convince the compiler that can be avoided by just jamming both sink & stream in the pubsub.

Reply:
You still need to have special logic for `PubSub`. `PubSub` needs to store pubsub messages while looking for a response but `SplitPubSubStream` needs to forward the response to a channel. And you can't do a delegation to the methods of the Sink because you need to poll stream to start the forward. You would have to spawn a task that would consume messages from the stream.

IMO I think this is clean enough for now, unless you have any ideas for how to handling the combination.

Reply:
PubSub needs to store pubsub messages while looking for a response but SplitPubSubStream needs to forward the response to a channel.

This sounds solvable to me, but I'm fine with continuing with the current design and with broad and robust testing, and then we can try after this PR to change the internal implementation without breaking the external behavior of the pubsub.